### PR TITLE
Added solution to Word Ladder

### PR DESCRIPTION
**Pull request by amemov**

### CODE REVIEW

**Comment by amemov** [comment]
Reviewed the code: addressed some small problems with types, refactored the structure, and reformatted the names to match Google's Format

File: Word Ladder/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    hdrs = ["solution.h"],
+    deps = [ 
+        "@abseil-cpp//absl/container:flat_hash_map",
+    ]
+
+)
+
+cc_test(
+    name = "solution_test",
+    size = "small",
```
Comment (kcc) [inline Word Ladder/BUILD.bazel]: redundant - drop

File: Word Ladder/MODULE.bazel.lock
Comment (kcc) [inline Word Ladder/MODULE.bazel.lock:1]: you probably don't want this file in the PR

File: Word Ladder/readme.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Solution to Leetcode [127 problem](https://leetcode.com/problems/word-ladder/) written in C++, following [Google format](https://google.github.io/styleguide/cppguide)
+
+To run, type the following in the terminal:
+```
+bazel test --cxxopt=-std=c++17 --test_output=all //:solution_test
```
Comment (kcc) [inline Word Ladder/readme.md:5]: Note: if you want to run a single test and see the output, you can use `bazel run`

File: Word Ladder/solution.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+#include <deque>
+#include <string>
+#include <vector>
+
+#include "absl/container/flat_hash_map.h"
+
+// This function return the number of words in the shortest transformation 
+// sequence from begin_word to end_word. Returns 0 if no such sequence.
+// Word Ladder Problem (LeetCode 127)
+int LadderLength(std::string begin_word, std::string end_word,
```
Comment (kcc) [inline Word Ladder/solution.h]: prefer to pass arguments as `const std::string &` or (better) `std::string_view`.
The comment does not explain the meaning of `word_list`, 
It it's a dictionary, it should be const.

File: Word Ladder/solution.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+  // Base case: end_word is NOT in word_list
+  if (find(word_list.begin(), word_list.end(), end_word) == word_list.end()) {
+    return 0;
+  }
+
+  // Create hashMap to store possible paths for every letter change
+  // for each word
+  absl::flat_hash_map<std::string, std::vector<std::string> > neighbors_map;
+  int word_len = begin_word.length();  // words are of equal length
```
Comment (kcc) [inline Word Ladder/solution.cpp]: if anything is a size/length, its type is size_t

Comment (kcc) [reply Word Ladder/solution.cpp]: use CHECK_EQ to verify that the two words are of equal length. 
Or return 0 if not.

File: Word Ladder/solution.cpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+  // Go through each word and create store possible patterns
```
Comment (kcc) [inline Word Ladder/solution.cpp]: typo?

File: Word Ladder/solution.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  // associated with a given word
+  // E.g neighbors_map[h*t] = {hot, hit}
+  for (std::string word_ : word_list) {
```
Comment (kcc) [inline Word Ladder/solution.cpp]: don't iterate using std::string: it creates a copy. 
use `for (auto & word: word_list)`

Also, check the variable naming

File: Word Ladder/solution.cpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    for (int i = 0; i < word_len; ++i) {
```
Comment (kcc) [inline Word Ladder/solution.cpp]: size_t

File: Word Ladder/solution.cpp [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      words_queue.pop_front();
+      
+      if (current_word == end_word) {
+        return shortest_path;
+      }
+
+      // Given combinations, see where we can go and only go through elements
+      // That we haven't added to visited_set yet
+      for (int j = 0; j < word_len; ++j) {
+        std::string pattern_ =
```
Comment (kcc) [inline Word Ladder/solution.cpp]: this pattern appears twice - move it to a separate function or a local lambda

File: Word Ladder/solution_test.cpp [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  std::string begin_word = "stone";
+  std::string end_word = "money";
+  std::vector<std::string> word_list = {
+      "stony", "store", "score", "scone", "shone", "phone", "phony", "money",
+      "stone", "stoke", "smoke", "smote", "motel", "model", "moody", "honey"};
+
+  EXPECT_EQ(LadderLength(begin_word, end_word, word_list), 0);
+}
+
+// Main function to run all tests
```
Comment (kcc) [inline Word Ladder/solution_test.cpp]: you have         "@googletest//:gtest_main", in deps, you don't need main() here

File: Word Ladder/solution.cpp [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  // for each word
+  absl::flat_hash_map<std::string, std::vector<std::string> > neighbors_map;
+  int word_len = begin_word.length();  // words are of equal length
+
+  // Go through each word and create store possible patterns
+  // associated with a given word
+  // E.g neighbors_map[h*t] = {hot, hit}
+  for (std::string word_ : word_list) {
+    for (int i = 0; i < word_len; ++i) {
+      std::string pattern_ =
```
Comment (kcc) [inline Word Ladder/solution.cpp]: I would write this code like this: 
   std::string pattern = word;
   pattern[i] = '*'

File: Word Ladder/solution.cpp [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+          word_.substr(0, i) + '*' + word_.substr(i + 1, word_len);
+      neighbors_map[pattern_].push_back(word_);
+    }
+  }
+
+  // Run BFS on neighbors_map. If we reached the word, return shortest_path
+  // If can't, while-loop will eventually run out of element in words_queue 
+  // and return 0
+  int shortest_path = 1;
+  std::set<std::string> visited_set;
```
Comment (kcc) [inline Word Ladder/solution.cpp]: you can also use flat_hash_set.
It's faster and has .contains()

File: Word Ladder/solution.cpp [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        return shortest_path;
+      }
+
+      // Given combinations, see where we can go and only go through elements
+      // That we haven't added to visited_set yet
+      for (int j = 0; j < word_len; ++j) {
+        std::string pattern_ =
+            current_word.substr(0, j) + '*' + current_word.substr(j + 1, word_len);
+        
+        for (std::string neighbor_word : neighbors_map[pattern_]) {
```
Comment (kcc) [inline Word Ladder/solution.cpp]: for (const auto & var :

**Review by kcc — COMMENTED** [review]
Yep, looks like a valid solution. Added a few style comments.

File: Word Ladder/solution.cpp (hunk 4 above)
Comment (amemov) [reply Word Ladder/solution.cpp]: Problem constraints gurantee us that the begin_word and end_word are of equal length. Same applies to words inside word_list vector - all words are guranteed to be of equal length, according to the problem definition on the LeetCode. In this case I don't think checking lengths would be necessary, nonetheless, it is a reasonable edge case.

File: Word Ladder/solution.cpp (hunk 11 above)
Comment (amemov) [reply Word Ladder/solution.cpp]: That's a good point about .contains().
I initially tried it with .find() and it showed 3 times slow-down over std::set, so I switched back to std::set. Now I tested flat_hash_set with .contains() and it works better

File: Word Ladder/BUILD.bazel [hunk 13]
```diff
@@ -4,13 +4,12 @@ cc_library(
     hdrs = ["solution.h"],
     deps = [ 
         "@abseil-cpp//absl/container:flat_hash_map",
+        "@abseil-cpp//absl/container:flat_hash_set",
+        "@com_github_google_glog//:glog",
```
Comment (kcc) [inline Word Ladder/BUILD.bazel]: glog is deprecated, use absl/log/log.g

File: Word Ladder/solution.cpp [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    }
+  }
+
+  // Run BFS on neighbors_map. If we reached the word, return shortest_path
+  // If can't, while-loop will eventually run out of element in words_queue
+  // and return 0
+  int shortest_path = 1;
+  absl::flat_hash_set<std::string> visited_set;
+  std::deque<std::string> words_queue;
+  words_queue.push_back(static_cast<std::string>(begin_word));
```
Comment (kcc) [inline Word Ladder/solution.cpp]: don't cast from string_view, create from it: std::string(begin_word)

**Review by kcc — COMMENTED** [review]
this is pretty good. If you want to do another problem like this, please try to not look at other existing solution at all, and come up with your own instead.

**Pull request merged**